File: appjava/src/main/java/com/byoutline/kickmaterial/utils/QueryParamsExtractor.java
package com.byoutline.kickmaterial.utils;

import timber.log.Timber;

import java.io.UnsupportedEncodingException;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLDecoder;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Helper methods fot extracting query parameters from url.
 * Created by Sebastian Kacprzak <sebastian.kacprzak at byoutline.com> on 30.03.15.
 */
public final class QueryParamsExtractor {

    private QueryParamsExtractor() {
    }

    public static Map<String, String> getQueryParams(String url) {
        try {
            URL u = new URL(url);
            return splitQuery(u);
        } catch (MalformedURLException | UnsupportedEncodingException e) {
            Timber.e("Failed to get query params from url: " + url, e);
            return Collections.emptyMap();
        }
    }

    public static Map<String, String> splitQuery(URL url) throws UnsupportedEncodingException {
        Map<String, String> queryPairs = new LinkedHashMap<>();
        String query = url.getQuery();
        String[] pairs = query.split("&");
        String charsetName = "UTF-8";
        for (String pair : pairs) {
            int idx = pair.indexOf("=");
            String key = URLDecoder.decode(pair.substring(0, idx), charsetName);
            String value = URLDecoder.decode(pair.substring(idx + 1), charsetName);
            queryPairs.put(key, value);
        }
        return queryPairs;
    }

}

File: app/src/main/java/com/byoutline/kickmaterial/utils/QueryParamsExtractor.kt
package com.byoutline.kickmaterial.utils

import timber.log.Timber
import java.io.UnsupportedEncodingException
import java.net.MalformedURLException
import java.net.URL
import java.net.URLDecoder
import java.util.*

/**
 * Helper methods fot extracting query parameters from url.
 * Created by Sebastian Kacprzak <sebastian.kacprzak at byoutline.com> on 30.03.15.
 */
object QueryParamsExtractor {

    fun getQueryParams(url: String): Map<String, String> {
        try {
            val u = URL(url)
            return splitQuery(u)
        } catch (e: MalformedURLException) {
            Timber.e("Failed to get query params from url: " + url, e)
            return emptyMap()
        } catch (e: UnsupportedEncodingException) {
            Timber.e("Failed to get query params from url: " + url, e)
            return emptyMap()
        }

    }

    @Throws(UnsupportedEncodingException::class)
    fun splitQuery(url: URL): Map<String, String> {
        val queryPairs = LinkedHashMap<String, String>()
        val query = url.query
        val pairs = query.split("&".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val charsetName = "UTF-8"
        for (pair in pairs) {
            val idx = pair.indexOf("=")
            val key = URLDecoder.decode(pair.substring(0, idx), charsetName)
            val value = URLDecoder.decode(pair.substring(idx + 1), charsetName)
            queryPairs.put(key, value)
        }
        return queryPairs
    }

}
